node/src/utils/median.rs

```diff
@@ -0,0 +1,55 @@
+use std::{
+    collections::BTreeMap,
+    ops::{Add, AddAssign},
+};
+
+pub(crate) fn weighted_median<T, I, Weight>(iterator: I) -> Option<T>
+where
+    T: Ord,
+    Weight: Default + Add<Weight, Output = Weight> + AddAssign<Weight> + PartialOrd + Copy,
+    I: IntoIterator<Item = (T, Weight)>,
+{
+    let (value_weights, total_weight): (_, Weight) = iterator.into_iter().fold(
+        (BTreeMap::<T, Weight>::new(), Default::default()),
+        |(mut weights, total), (value, weight)| {
+            *weights.entry(value).or_default() += weight;
+            (weights, total + weight)
+        },
+    );
+    let mut counted: Weight = Default::default();
+    let mut result = None;
+    for (value, weight) in value_weights {
+        result = Some(value);
+        counted += weight;
+        if counted + counted >= total_weight {
```

Comment:
Just to make sure we don't overflow (as long as the total doesn't overflow, which we can assume):
```suggestion
        if counted.saturating_add(counted) >= total_weight {
```
Or:
```rust
        if counted.saturating_mul(2) >= total_weight {
```
Or maybe:
```rust
        if counted > total_weight / 2 {
```

Reply:
I went with `counted + counted` to avoid bringing more traits into scope (and to avoid specifying which number type I'm using when multiplying / dividing!). Is `saturating_add` also working with types that just implement `Add`? I would be surprised, TBH...
And if we go with division, which number type should I use? I'll need to specify that `Weight: Div<something>` (would be nice to be able to specify `Div<N> where N: Number`, but is there a trait implemented by all number types already?).

Reply:
That's true, you'd need to implement `saturating_add` or `saturating_mul` or `Div<u64>` or `Div<u8>` or something for `Weight`. Not sure what's best.

Reply:
But _if_ we keep using `counted + counted`, we need to somehow ensure that the total weight of our validators never exceeds `u64::MAX / 2`.

Reply:
Note that `Weight` in `weighted_median` is a generic type parameter, not the `Weight` type from Highway. So just implementing these for `Weight` wouldn't help :grimacing: 
Anyway, I'll just add a constraint `Weight: Div<u64>` and work with that.